### PR TITLE
feat: add account recovery functionality

### PR DESCRIPTION
**Pull request by keyvankhademi**

### CODE REVIEW

File: staking/programs/staking/src/context.rs [hunk 1]
```diff
@@ -405,3 +405,38 @@ pub struct AdvanceClock<'info> {
...
+    pub stake_account_metadata: Account<'info, stake_account::StakeAccountMetadataV2>,
+
+    #[account(
+        mut,
+        seeds = [VOTER_RECORD_SEED.as_bytes(), stake_account_positions.key().as_ref()],
+        bump = stake_account_metadata.voter_bump
+    )]
+    pub voter_record: Account<'info, voter_weight_record::VoterWeightRecord>,
+
+    #[account(mut, seeds = [CONFIG_SEED.as_bytes()], bump = config.bump)]
```
Comment (guibescos) [inline staking/programs/staking/src/context.rs:440]: This doesn't need to be `mut`

File: staking/tests/recouver_account.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
+import {
```
Comment (guibescos) [inline staking/tests/recouver_account.ts:1]: Typo in the name of this file : `recouver` -> `recover`

File: staking/tests/recouver_account.ts [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    await governanceConnection.program.provider.sendAndConfirm(
+      recoverAccountTransaction
+    );
+
+    const fixedAccount = await governanceConnection.loadStakeAccount(
+      badStakeAccountAddress
+    );
+
+    assert.equal(
+      fixedAccount.stakeAccountPositionsJs.owner.toString(),
```
Comment (guibescos) [inline staking/tests/recouver_account.ts:188]: Perhaps you want to check the owner field in `stakeAccountMetadata` too

File: staking/tests/recouver_account.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        provider.wallet.publicKey
+      );
+    const recoverAccountTransaction = new Transaction();
+    recoverAccountTransaction.instructions.push(recoverAccountInstruction);
+
+    await expectFailApi(
+      aliceConnection.program.provider.sendAndConfirm(
+        recoverAccountTransaction
+      ),
+      "Signature verification failed"
```
Comment (guibescos) [inline staking/tests/recouver_account.ts:258]: This is not the error that we want

Comment (keyvankhademi) [reply staking/tests/recouver_account.ts:258]: Fixed

File: staking/tests/recouver_account.ts (hunk 3 above)
Comment (keyvankhademi) [reply staking/tests/recouver_account.ts:188]: Done

File: staking/tests/recouver_account.ts (hunk 2 above)
Comment (keyvankhademi) [reply staking/tests/recouver_account.ts:1]: Fixed

File: staking/programs/staking/src/context.rs (hunk 1 above)
Comment (keyvankhademi) [reply staking/programs/staking/src/context.rs:440]: Fixed

File: staking/programs/staking/src/lib.rs [hunk 5]
```diff
@@ -706,4 +706,15 @@ pub mod staking {
             Some(ctx.accounts.config.agreement_hash);
         Ok(())
     }
+
+    pub fn recover_account(ctx: Context<RecoverAccount>) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (keyvankhademi) [reply staking/programs/staking/src/lib.rs:716]: Done

File: staking/programs/staking/src/lib.rs [hunk 6]
```diff
@@ -706,4 +706,21 @@ pub mod staking {
...
     }
+
+    /** Recovers a user's `stake account` ownership by transferring ownership
+     * from a token account to the `owner` of that token account.
+     *
+     * This functionality addresses the scenario where a user mistakenly
+     * created a stake account using their token account address as the owner.
+     */
+    pub fn recover_account(ctx: Context<RecoverAccount>) -> Result<()> {
+        let new_owner = ctx.accounts.payer_token_account.owner;
```
Comment (guibescos) [inline staking/programs/staking/src/lib.rs:717]: Can you add a check like the one in line 627 to assert this account has no staked tokens?

File: staking/target/idl/staking.json [hunk 7]
```diff
@@ -1198,6 +1198,79 @@
           }
         }
       ]
+    },
+    {
```
Comment (guibescos) [inline staking/target/idl/staking.json:1202]: I think the idls are out of date and it's making CI fail

File: staking/tests/recover_account.ts [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+import { PYTH_DECIMALS, PythBalance, StakeConnection } from "../app";
+import NodeWallet from "@coral-xyz/anchor/dist/cjs/nodewallet";
+
+// When DEBUG is turned on, we turn preflight transaction checking off
+// That way failed transactions show up in the explorer, which makes them
+// easier to debug.
+const DEBUG = true;
+const portNumber = getPortNumber(path.basename(__filename));
+
+describe("config", async () => {
```
Comment (guibescos) [inline staking/tests/recover_account.ts:35]: "config" seems ambiguous here

File: staking/tests/recover_account.ts [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    );
+
+    assert.equal(
+      fixedAccount.stakeAccountPositionsJs.owner.toString(),
+      newOwner.publicKey.toString()
+    );
+    assert.equal(
+      fixedAccount.stakeAccountMetadata.owner.toString(),
+      newOwner.publicKey.toString()
+    );
```
Comment (guibescos) [inline staking/tests/recover_account.ts:179]: Can we add these asserts before too, to assert the owner is newOwnerAta before we call the instruction

File: staking/programs/staking/src/lib.rs (hunk 6 above)
Comment (guibescos) [reply staking/programs/staking/src/lib.rs:717]: Transferring accounts with staked tokens might lead to double voting

File: staking/programs/staking/Cargo.toml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "pyth-staking-program"
-version = "1.0.0"
+version = "1.1.0"
```
[the remark this answers is not in the document]
Comment (keyvankhademi) [reply staking/programs/staking/Cargo.toml:3]: Done

File: staking/tests/recover_account.ts (hunk 8 above)
Comment (keyvankhademi) [reply staking/tests/recover_account.ts:35]: Fixed

File: staking/target/idl/staking.json (hunk 7 above)
Comment (keyvankhademi) [reply staking/target/idl/staking.json:1202]: Fixed

File: staking/tests/recover_account.ts (hunk 9 above)
Comment (keyvankhademi) [reply staking/tests/recover_account.ts:179]: Done

File: staking/programs/staking/src/lib.rs (hunk 6 above)
Comment (keyvankhademi) [reply staking/programs/staking/src/lib.rs:724]: Done

File: staking/programs/staking/src/lib.rs [hunk 11]
```diff
@@ -706,4 +706,28 @@ pub mod staking {
...
+     *
+     * This functionality addresses the scenario where a user mistakenly
+     * created a stake account using their token account address as the owner.
+     */
+    pub fn recover_account(ctx: Context<RecoverAccount>) -> Result<()> {
+        // Check that there aren't any positions (i.e., staked tokens) in the account.
+        // Transferring accounts with staked tokens might lead to double voting
+        require!(
+            ctx.accounts.stake_account_metadata.next_index == 0,
+            ErrorCode::SplitWithStake
```
Comment (guibescos) [inline staking/programs/staking/src/lib.rs:721]: It's weird you created a new error `RecoverWithStake` but you aren't using it here.

**Review by guibescos — APPROVED** [review]
One last nit

File: staking/programs/staking/src/lib.rs (hunk 11 above)
Comment (keyvankhademi) [reply staking/programs/staking/src/lib.rs:721]: I fixed this i https://github.com/pyth-network/governance/pull/426

**Pull request merged**